### DISCUSSION TITLE
WebsocketException reason is not visible to the client

### DISCUSSION THREAD

**Discussion by PawelRoman** [fastapi/fastapi#6076] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi import WebSocketException, status

app = FastAPI()

@app.websocket("/ws")
async def websocket_endpoint():
   raise WebSocketException(code=status.WS_1008_POLICY_VIOLATION, reason="I want you to know this")
```

### Description

The reason code and text seem to be unavailable for the client. The only way to get them is to have websocket client written in python, import starlette's WebSocketDisconnect and catch the exception manually. 

When making calls from other clients, e.g. from Postman, the code and reason are just not available. It would be nice if the server returned code and reason as handshake response headers e.g. X-Code and X-Reason or something. Not sure if that makes sense?

### Wanted Solution

Have code and reason from WebSocketException be surfaced as HTTP response headers

### Wanted Code

```python
Put any non-python websocket client code here, connect to websocket and try to get code and reason on failed connection.
```

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.88.0

### Python Version

3.11

**Comment by Kludex** [comment, marked as answer]
No, it doesn't make sense. The handshake already happened when you raise the exception.

Have you tried another WebSocket client?

**Comment by VolVox99** [comment]
Following up on this. It's still an issue. @Kludex

**Comment by bhumkong** [comment]
As you can see on [ASGI page](https://asgi.readthedocs.io/en/latest/extensions.html),

> Websocket connections start with the client sending a HTTP request containing the appropriate upgrade headers. On receipt of this request a server can choose to either upgrade the connection or respond with an HTTP response (denying the upgrade). **The core ASGI specification does not allow for any control over the denial response, instead specifying that the HTTP status code 403 should be returned**

They are adding _Websocket Denial Response_ extension to support custom HTTP responses in this case. You can use this extension when it becomes available in FastAPI or you can accept the ws connection before raising `WebSocketException`.
